extensions/amp-video/0.1/test-e2e/test-amp-video-flexible-bitrate.js

```diff
@@ -0,0 +1,232 @@
+      await expect(
+        await controller.getElementProperty(video4El, 'currentSrc')
+      ).to.equal('');
+    });
+
+    /**
+     * Tests for downgrades on active video
+     */
+
+    it('when a video buffers, the video changes to a lower quality source', async () => {
+      await forceEventOnVideo(VIDEO_EVENTS.UNLOAD, 1);
+
+      const video1El = await controller.findElement('#video1 video');
+
+      await forceEventOnVideo(VIDEO_EVENTS.DOWNGRADE, 1);
+
+      await expect(
+        await controller.getElementProperty(video1El, 'currentSrc')
+      ).contains('#med');
+    });
+
+    it('when a video buffers, the low quality source starts playing at the same place the previous source was stopped at', async () => {
+      const video1El = await controller.findElement('#video1 video');
+
+      await sleep(100);
+
+      const currentTime = await controller.getElementProperty(
+        video1El,
+        'currentTime'
+      );
+
+      await forceEventOnVideo(VIDEO_EVENTS.UNLOAD, 1);
+      await forceEventOnVideo(VIDEO_EVENTS.DOWNGRADE, 1);
+
+      await expect(
+        await controller.getElementProperty(video1El, 'currentTime')
+      ).to.be.gte(currentTime);
+    });
+
+    it("when a video buffers many times, it doesn't keep reloading", async () => {
+      await forceEventOnVideo(VIDEO_EVENTS.UNLOAD, 1);
+      await forceEventOnVideo(VIDEO_EVENTS.DOWNGRADE, 1);
+      await forceEventOnVideo(VIDEO_EVENTS.DOWNGRADE, 1);
+      await forceEventOnVideo(VIDEO_EVENTS.DOWNGRADE, 1);
+
+      const video1El = await controller.findElement(
+        '#video1 video.i-amphtml-pool-video'
+      );
+
+      await expect(
+        await controller.getElementProperty(video1El, 'currentSrc')
+      ).contains('#low');
+      await expect(await controller.getElementProperty(video1El, 'paused')).is
+        .false;
+    });
+
+    /**
+     * Tests for downgrades on inactive videos
+     */
+
+    it('when a video buffers, other videos that are not already loaded lower the quality source', async () => {
+      await forceEventOnVideo(VIDEO_EVENTS.UNLOAD, 2);
+      await forceEventOnVideo(VIDEO_EVENTS.DOWNGRADE, 1);
+      await controller.click(story);
+      await controller.findElement('amp-story-page#page-2[active]');
+
+      const video2El = await controller.findElement('#video2 video');
+
+      await expect(
+        await controller.getElementProperty(video2El, 'currentSrc')
+      ).contains('#med');
+    });
+
+    it("when a video buffers, other videos that are already loaded don't lower the quality source", async () => {
+      await forceEventOnVideo(VIDEO_EVENTS.LOAD, 2);
+      await forceEventOnVideo(VIDEO_EVENTS.DOWNGRADE, 1);
+
+      const video4El = await controller.findElement(
+        '#video2 video.i-amphtml-pool-video'
+      );
+
+      await expect(
+        await controller.getElementProperty(video4El, 'currentSrc')
+      ).contains('#high');
+    });
+
+    /**
+     * Tests on source generation
```

Comment:
You forgot a test to check that one cached source generates 3 sources with extra quality URL params, and a data-bitrate attribute